src/graph.rs

```diff
@@ -3426,19 +3426,46 @@ impl<'a, 'graph> Builder<'a, 'graph> {
     package_req: &PackageReq,
   ) -> Option<Version> {
     // try to find in the list of existing versions first
-    if let Some(existing_versions) =
+    let versions = if let Some(existing_versions) =
       self.graph.packages.versions_by_name(&package_req.name)
     {
-      if let Some(version) = resolve_version(
-        &package_req.version_req,
-        existing_versions.iter().map(|nv| &nv.version),
-      ) {
+      existing_versions
+        .iter()
+        .map(|nv| &nv.version)
+        .collect::<Vec<_>>()
+    } else {
+      package_info.versions.keys().collect::<Vec<_>>()
+    };
+    let is_yanked = |v| {
+      package_info
+        .versions
+        .get(v)
+        .map(|i| i.yanked)
+        .unwrap_or(false)
+    };
+    let mut resolved_yanked_version = None;
+    if let Some(version) =
+      resolve_version(&package_req.version_req, versions.iter().copied())
+    {
+      if !is_yanked(version) {
         return Some(version.clone());
       }
+      resolved_yanked_version = Some(version.clone());
     }
-    // now try in the package info
-    resolve_version(&package_req.version_req, package_info.versions.keys())
-      .cloned()
+    let unyanked_versions = versions.iter().copied().filter(|v| !is_yanked(v));
+    if let Some(version) =
+      resolve_version(&package_req.version_req, unyanked_versions)
+    {
+      return Some(version.clone());
+    }
+    if let Some(version) = resolved_yanked_version {
+      self.graph.packages.add_used_yanked_package(PackageNv {
+        name: package_req.name.clone(),
+        version: version.clone(),
+      });
+      return Some(version);
+    };
+    None
   }
```

Comment:
I think we could reduce the hashmap lookups and cloning by doing something like this?

```rs
  fn resolve_jsr_version(
    &mut self,
    package_info: &JsrPackageInfo,
    package_req: &PackageReq,
  ) -> Option<Version> {
    // 1. try to resolve with the list of existing versions
    if let Some(existing_versions) =
      self.graph.packages.versions_by_name(&package_req.name)
    {
      if let Some(version) = resolve_version(
        &package_req.version_req,
        existing_versions.iter().map(|nv| &nv.version),
      ) {
        let is_yanked = package_info
          .versions
          .get(version)
          .map(|i| i.yanked)
          .unwrap_or(false);
        let version = version.clone();
        if is_yanked {
          self.graph.packages.add_used_yanked_package(PackageNv {
            name: package_req.name.clone(),
            version: version.clone(),
          });
        }
        return Some(version);
      }
    }

    // 2. attempt to resolve with the unyanked versions
    let unyanked_versions =
      package_info.versions.iter().filter_map(|(v, i)| {
        if !i.yanked {
          Some(v)
        } else {
          None
        }
      });
    if let Some(version) =
      resolve_version(&package_req.version_req, unyanked_versions)
    {
      return Some(version.clone());
    }

    // 3. attempt to resolve with the the yanked versions
    let yanked_versions = package_info.versions.iter().filter_map(|(v, i)| {
      if i.yanked {
        Some(v)
      } else {
        None
      }
    });
    if let Some(version) =
      resolve_version(&package_req.version_req, yanked_versions)
    {
      self.graph.packages.add_used_yanked_package(PackageNv {
        name: package_req.name.clone(),
        version: version.clone(),
      });
      return Some(version.clone());
    }

    None
  }
```